File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/StatisticsSample/src/com/hp/workpath/sample/statisticsample/fragment/TestJobFragment.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.statisticsample.fragment;

import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.text.method.ScrollingMovementMethod;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.hp.workpath.api.CapabilitiesExceededException;
import com.hp.workpath.api.Result;
import com.hp.workpath.api.job.JobInfo;
import com.hp.workpath.api.job.JobService;
import com.hp.workpath.api.job.JobletAttributes;
import com.hp.workpath.api.printer.PrintAttributes;
import com.hp.workpath.api.printer.PrintAttributesCaps;
import com.hp.workpath.api.scanner.ScanAttributes;
import com.hp.workpath.api.scanner.ScanAttributesCaps;
import com.hp.workpath.api.statistics.StatisticsJobData;
import com.hp.workpath.sample.statisticsample.Logger;
import com.hp.workpath.sample.statisticsample.MainActivity;
import com.hp.workpath.sample.statisticsample.R;
import com.hp.workpath.sample.statisticsample.task.LastJobInfoTask;
import com.hp.workpath.sample.statisticsample.task.RequestPrintTask;
import com.hp.workpath.sample.statisticsample.task.RequestScanTask;

import java.io.File;
import java.util.List;

public class TestJobFragment extends Fragment implements View.OnClickListener {

    Button mGetLastJobInfoButton;
    Button mPrintButton;
    Button mScanButton;
    TextView mJobInfoTextView;
    TextView mStatisticInfoTextView;
    ProgressBar mProgressBar;

    private JobObserver mJobObserver = null;
    private String mRid = null;
    private String mJobId = null;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mJobObserver = new JobObserver(new Handler());
        mJobObserver.register(getContext());
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_test_job, container, false);
        mGetLastJobInfoButton = view.findViewById(R.id.getLastJobInfoButton);
        mGetLastJobInfoButton.setOnClickListener(this);
        mPrintButton = view.findViewById(R.id.printButton);
        mPrintButton.setOnClickListener(this);
        mScanButton = view.findViewById(R.id.scanButton);
        mScanButton.setOnClickListener(this);
        mJobInfoTextView = view.findViewById(R.id.jobInfoTextView);
        mJobInfoTextView.setMovementMethod(new ScrollingMovementMethod());
        mStatisticInfoTextView = view.findViewById(R.id.statisticInfoTextView);
        mStatisticInfoTextView.setMovementMethod(new ScrollingMovementMethod());
        mProgressBar = view.findViewById(R.id.progressBar);
        return view;
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mJobObserver.unregister(getContext());
    }

    @Override
    public void onClick(View v) {
        mProgressBar.setVisibility(View.VISIBLE);
        if (v == mGetLastJobInfoButton) {
            new LastJobInfoTask(getContext(), statisticsInterface).taskExecute();
        } else if (v == mPrintButton) {
            PrintAttributesCaps printCaps = ((MainActivity) getActivity()).getPrintCaps();
            if (printCaps == null) {
                Logger.showResult(getActivity(), "Print is not supported");
                mProgressBar.setVisibility(View.GONE);
                return;
            }
            enableButton(false);
            cleanJobInfo();
            try {
                String path = getActivity().getFilesDir().getPath() + "/test_page.pdf";
                PrintAttributes attributes = new PrintAttributes.PrintFromStorageBuilder(Uri.fromFile(new File(path)))
                        .build(printCaps);
                new RequestPrintTask(this, attributes).taskExecute();
            } catch (CapabilitiesExceededException e) {
                enableButton(true);
                mProgressBar.setVisibility(View.GONE);
                Logger.showResult(getActivity(), "PrintAttributes.PrintFromStorageBuilder" + e.getMessage());
            }
        } else if (v == mScanButton) {
            ScanAttributesCaps scanCaps = ((MainActivity) getActivity()).getScanCaps();
            if (scanCaps == null) {
                Logger.showResult(getActivity(), "Scan is not supported");
                mProgressBar.setVisibility(View.GONE);
                return;
            }
            enableButton(false);
            cleanJobInfo();
            try {
                ScanAttributes.MeBuilder attributesBuilder = new ScanAttributes.MeBuilder();
                attributesBuilder.setDocumentFormat(ScanAttributes.DocumentFormat.PDF);
                attributesBuilder.setJobAssemblyMode(ScanAttributes.JobAssemblyMode.OFF);
                new RequestScanTask(this, attributesBuilder.build(scanCaps)).taskExecute();
            } catch (CapabilitiesExceededException e) {
                enableButton(true);
                mProgressBar.setVisibility(View.GONE);
                Logger.showResult(getActivity(), "ScanAttributes.MeBuilder" + e.getMessage());
            }
        }
    }

    private void cleanJobInfo() {
        mRid = null;
        mJobId = null;
    }

    private void enableButton(boolean enable) {
        if (mPrintButton != null) {
            mPrintButton.setEnabled(enable);
        }
        if (mScanButton != null) {
            mScanButton.setEnabled(enable);
        }
    }

    public void setRid(String rid) {
        mRid = rid;
    }

    ResponseInterface statisticsInterface = new ResponseInterface() {
        @Override
        public void success(List<StatisticsJobData> info) {
            mProgressBar.setVisibility(View.GONE);
            if (info != null && info.size() > 0) {
                StatisticsJobData data = info.get(0);
                if (TextUtils.isEmpty(data.getJobId())) {
                    mStatisticInfoTextView.setText("There is no statistic information");
                } else {
                    Gson gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
                    mStatisticInfoTextView.setText(gson.toJson(data));
                }
            } else {
                mStatisticInfoTextView.setText("There is no statistic information");
            }
        }

        @Override
        public void failure(String msg, Result result) {
            mProgressBar.setVisibility(View.GONE);
            Logger.showResult(getActivity(), msg, result);
        }
    };

    private class JobObserver extends JobService.AbstractJobletObserver {

        public JobObserver(final Handler handler) {
            super(handler);
        }

        @Override
        public void onComplete(String rid, JobInfo jobInfo) {
            mProgressBar.setVisibility(View.GONE);
            Gson gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create();
            mJobInfoTextView.setText(gson.toJson(jobInfo));
            enableButton(true);
        }

        @Override
        public void onProgress(String rid, JobInfo jobInfo) {
            Log.d(MainActivity.TAG, "Received onProgress for rid " + rid);
            Log.d(MainActivity.TAG, "Received onProgress jobInfo " + jobInfo);
            if (rid.equals(mRid)) {
                if (mJobId == null) {
                    if (jobInfo.getJobId() != null) {
                        mJobId = jobInfo.getJobId();
                        Log.d(MainActivity.TAG, "Received jobID as " + mJobId);
                        final JobletAttributes taskAttributes =
                                new JobletAttributes.Builder().setShowUi(true).build();

                        final String jrid = JobService.monitorJobInForeground(getActivity(), mJobId,
                                taskAttributes, null);
                        Log.d(MainActivity.TAG, "MonitorJob request: " + jrid);
                    }
                }
            }
        }

        @Override
        public void onFail(String rid, Result result) {
            Log.e(MainActivity.TAG, "Received onFail for rid " + rid + ", " + result);
            mProgressBar.setVisibility(View.GONE);
            mJobInfoTextView.setText("Received onFail for rid " + rid + ", " + Logger.build(result));
            enableButton(true);
        }

        @Override
        public void onCancel(String rid) {
            Log.d(MainActivity.TAG, "Received onCancel for rid " + rid);
            mProgressBar.setVisibility(View.GONE);
            mJobInfoTextView.setText("Received onCancel for rid " + rid);
            enableButton(true);
        }
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/StatisticsSample/src/com/hp/workpath/sample/statisticsample/fragment/TestJobFragment.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.statisticsample.fragment

import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.text.TextUtils
import android.text.method.ScrollingMovementMethod
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import com.google.gson.GsonBuilder
import com.hp.workpath.api.CapabilitiesExceededException
import com.hp.workpath.api.Result
import com.hp.workpath.api.job.JobInfo
import com.hp.workpath.api.job.JobService
import com.hp.workpath.api.job.JobService.AbstractJobletObserver
import com.hp.workpath.api.job.JobletAttributes
import com.hp.workpath.api.printer.PrintAttributes.PrintFromStorageBuilder
import com.hp.workpath.api.scanner.ScanAttributes
import com.hp.workpath.api.scanner.ScanAttributes.MeBuilder
import com.hp.workpath.api.statistics.StatisticsJobData
import com.hp.workpath.sample.statisticsample.Logger
import com.hp.workpath.sample.statisticsample.Logger.build
import com.hp.workpath.sample.statisticsample.MainActivity
import com.hp.workpath.sample.statisticsample.databinding.FragmentTestJobBinding
import com.hp.workpath.sample.statisticsample.task.LastJobInfoTask
import com.hp.workpath.sample.statisticsample.task.RequestPrintTask
import com.hp.workpath.sample.statisticsample.task.RequestScanTask
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import java.io.File

class TestJobFragment : Fragment(), View.OnClickListener {

    private lateinit var mJobObserver: JobObserver
    private var mRid: String? = null
    private var mJobId: String? = null
    private var mBindingFragment: FragmentTestJobBinding? = null
    private val mBindingTestJobFragment get() = mBindingFragment!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mJobObserver = JobObserver(Handler(Looper.getMainLooper()))
        context?.let { mJobObserver.register(it) }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        mBindingFragment = FragmentTestJobBinding.inflate(inflater, container, false)
        return mBindingTestJobFragment.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mBindingTestJobFragment.statisticInfoTextView.movementMethod = ScrollingMovementMethod()
        mBindingTestJobFragment.jobInfoTextView.movementMethod = ScrollingMovementMethod()
        mBindingTestJobFragment.getLastJobInfoButton.setOnClickListener(this)
        mBindingTestJobFragment.printButton.setOnClickListener(this)
        mBindingTestJobFragment.scanButton.setOnClickListener(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        context?.let { mJobObserver.unregister(it) }
    }

    override fun onClick(v: View) {
        mBindingTestJobFragment.progressBar.visibility = View.VISIBLE
        when {
            v === mBindingTestJobFragment.getLastJobInfoButton -> {
                viewLifecycleOwner.lifecycleScope.launch(Dispatchers.Default) {
                    context?.let { LastJobInfoTask(it, statisticsInterface).execute() }
                }
            }
            v === mBindingTestJobFragment.printButton -> {
                val printCaps = (activity as MainActivity).getPrintCaps()
                if (printCaps == null) {
                    Logger.showResult(activity, "Print is not supported")
                    mBindingTestJobFragment.progressBar.visibility = View.GONE
                    return
                }
                enableButton(false)
                cleanJobInfo()
                try {
                    val path = activity?.filesDir?.path + "/test_page.pdf"
                    val attributes = PrintFromStorageBuilder(Uri.fromFile(File(path))).build(printCaps)
                    viewLifecycleOwner.lifecycleScope.launch(Dispatchers.Default) {
                        RequestPrintTask(this@TestJobFragment, attributes).execute()
                    }
                } catch (e: CapabilitiesExceededException) {
                    mBindingTestJobFragment.progressBar.visibility = View.GONE
                    enableButton(true)
                    Logger.showResult(
                        activity,
                        "PrintAttributes.PrintFromStorageBuilder" + e.message
                    )
                }
            }
            v === mBindingTestJobFragment.scanButton -> {
                val scanCaps = (activity as MainActivity).getScanCaps()
                if (scanCaps == null) {
                    Logger.showResult(activity, "Scan is not supported")
                    mBindingTestJobFragment.progressBar.visibility = View.GONE
                    return
                }
                enableButton(false)
                cleanJobInfo()
                try {
                    val attributesBuilder = MeBuilder()
                    attributesBuilder.setDocumentFormat(ScanAttributes.DocumentFormat.PDF)
                    attributesBuilder.setJobAssemblyMode(ScanAttributes.JobAssemblyMode.OFF)
                    viewLifecycleOwner.lifecycleScope.launch(Dispatchers.Default) {
                        RequestScanTask(this@TestJobFragment, attributesBuilder.build(scanCaps)).execute()
                    }
                } catch (e: CapabilitiesExceededException) {
                    mBindingTestJobFragment.progressBar.visibility = View.GONE
                    enableButton(true)
                    Logger.showResult(activity, "ScanAttributes.MeBuilder" + e.message)
                }
            }
        }
    }

    private fun cleanJobInfo() {
        mRid = null
        mJobId = null
    }

    private fun enableButton(enable: Boolean) {
        mBindingTestJobFragment.printButton.isEnabled = enable
        mBindingTestJobFragment.scanButton.isEnabled = enable
    }

    fun setRid(rid: String?) {
        mRid = rid
    }

    private var statisticsInterface: ResponseInterface = object : ResponseInterface {
        override fun success(info: List<StatisticsJobData?>) {
            mBindingTestJobFragment.progressBar.visibility = View.GONE
            if (info.isNotEmpty()) {
                val data = info[0]
                if (TextUtils.isEmpty(data?.jobId)) {
                    mBindingTestJobFragment.statisticInfoTextView.text =
                        "There is no statistic information"
                } else {
                    val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
                    mBindingTestJobFragment.statisticInfoTextView.text = gson.toJson(data)
                }
            } else {
                mBindingTestJobFragment.statisticInfoTextView.text =
                    "There is no statistic information"
            }
        }

        override fun failure(msg: String?, result: Result?) {
            mBindingTestJobFragment.progressBar.visibility = View.GONE
            Logger.showResult(activity, msg, result)
        }
    }

    private inner class JobObserver(handler: Handler?) : AbstractJobletObserver(handler) {
        override fun onComplete(rid: String, jobInfo: JobInfo) {
            mBindingTestJobFragment.progressBar.visibility = View.GONE
            val gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
            mBindingTestJobFragment.jobInfoTextView.text = gson.toJson(jobInfo)
            enableButton(true)
        }

        override fun onProgress(rid: String, jobInfo: JobInfo) {
            Log.d(MainActivity.TAG, "Received onProgress for rid $rid")
            Log.d(MainActivity.TAG, "Received onProgress jobInfo $jobInfo")
            if (rid == mRid) {
                if (mJobId == null) {
                    if (jobInfo.jobId != null) {
                        mJobId = jobInfo.jobId
                        Log.d(MainActivity.TAG, "Received jobID as $mJobId")
                        val taskAttributes = JobletAttributes.Builder().setShowUi(true).build()
                        val jrid = activity?.let {
                            JobService.monitorJobInForeground(
                                it,
                                jobInfo.jobId,
                                taskAttributes,
                                null
                            )
                        }
                        Log.d(MainActivity.TAG, "MonitorJob request: $jrid")
                    }
                }
            }
        }

        override fun onFail(rid: String, result: Result) {
            Log.e(MainActivity.TAG, "Received onFail for rid $rid, $result")
            mBindingTestJobFragment.progressBar.visibility = View.GONE
            mBindingTestJobFragment.jobInfoTextView.text =
                "Received onFail for rid " + rid + ", " + build(result)
            enableButton(true)
        }

        override fun onCancel(rid: String) {
            Log.d(MainActivity.TAG, "Received onCancel for rid $rid")
            mBindingTestJobFragment.progressBar.visibility = View.GONE
            mBindingTestJobFragment.jobInfoTextView.text = "Received onCancel for rid $rid"
            enableButton(true)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mBindingFragment = null
    }
}